In [2]:
import pandas as pd

df = pd.read_csv("messy_sales_data.csv")

print("Dataset loaded successfully!")
print(df.shape)
df.head()

Dataset loaded successfully!
(25, 11)


,Customer_ID,Name,Gender,Age,Join_Date,City,Product,Quantity,Unit_Price,Total_Sales,Email
0,C001,John Smith,Male,25.0,2023-01-15,Chennai,Laptop,2.0,55000,110000.0,john@gmail.com
1,C002,Priya Sharma,female,32.0,15/02/2023,Chennai,Phone,1.0,25000,25000.0,priya@gmail.com
2,C003,Rahul Kumar,M,28.0,2023/03/10,Bangalore,Tablet,3.0,18000,54000.0,rahul@gmail.com
3,C004,Anita Das,F,45.0,10-04-2023,Mumbai,Laptop,1.0,60000,60000.0,anita@gmail.com
4,C005,Arjun Patel,male,52.0,2023-05-20,Delhi,Phone,2.0,30000,60000.0,arjun@gmail.com


In [3]:
from google.colab import files
import pandas as pd
import numpy as np

# Upload CSV
uploaded = files.upload()

# Load dataset
df = pd.read_csv("messy_sales_data.csv")

# Display basic information
print("Dataset Shape:", df.shape)
df.head()

Saving messy_sales_data.csv to messy_sales_data (1).csv
Dataset Shape: (25, 11)


,Customer_ID,Name,Gender,Age,Join_Date,City,Product,Quantity,Unit_Price,Total_Sales,Email
0,C001,John Smith,Male,25.0,2023-01-15,Chennai,Laptop,2.0,55000,110000.0,john@gmail.com
1,C002,Priya Sharma,female,32.0,15/02/2023,Chennai,Phone,1.0,25000,25000.0,priya@gmail.com
2,C003,Rahul Kumar,M,28.0,2023/03/10,Bangalore,Tablet,3.0,18000,54000.0,rahul@gmail.com
3,C004,Anita Das,F,45.0,10-04-2023,Mumbai,Laptop,1.0,60000,60000.0,anita@gmail.com
4,C005,Arjun Patel,male,52.0,2023-05-20,Delhi,Phone,2.0,30000,60000.0,arjun@gmail.com


In [4]:
df = pd.read_csv("messy_sales_data.csv")

print("Dataset loaded successfully.")
print("Shape:", df.shape)

df.head()

Dataset loaded successfully.
Shape: (25, 11)


,Customer_ID,Name,Gender,Age,Join_Date,City,Product,Quantity,Unit_Price,Total_Sales,Email
0,C001,John Smith,Male,25.0,2023-01-15,Chennai,Laptop,2.0,55000,110000.0,john@gmail.com
1,C002,Priya Sharma,female,32.0,15/02/2023,Chennai,Phone,1.0,25000,25000.0,priya@gmail.com
2,C003,Rahul Kumar,M,28.0,2023/03/10,Bangalore,Tablet,3.0,18000,54000.0,rahul@gmail.com
3,C004,Anita Das,F,45.0,10-04-2023,Mumbai,Laptop,1.0,60000,60000.0,anita@gmail.com
4,C005,Arjun Patel,male,52.0,2023-05-20,Delhi,Phone,2.0,30000,60000.0,arjun@gmail.com


In [5]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

Number of rows: 25
Number of columns: 11

Column names:
['Customer_ID', 'Name', 'Gender', 'Age', 'Join_Date', 'City', 'Product', 'Quantity', 'Unit_Price', 'Total_Sales', 'Email']

Data types:
Customer_ID     object
Name            object
Gender          object
Age            float64
Join_Date       object
City            object
Product         object
Quantity       float64
Unit_Price       int64
Total_Sales    float64
Email           object
dtype: object


In [6]:
null_report = pd.DataFrame({
    "Null_Count": df.isnull().sum(),
    "Null_Percentage": (df.isnull().sum() / len(df) * 100).round(2)
})

null_report

,Null_Count,Null_Percentage
Customer_ID,0,0.0
Name,0,0.0
Gender,0,0.0
Age,1,4.0
Join_Date,0,0.0
City,0,0.0
Product,0,0.0
Quantity,1,4.0
Unit_Price,0,0.0
Total_Sales,1,4.0


In [7]:
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 0


In [8]:
df.describe(include="all")

,Customer_ID,Name,Gender,Age,Join_Date,City,Product,Quantity,Unit_Price,Total_Sales,Email
count,25,25,25,24.000000,25,25,25,24.000000,25.000000,24.000000,25
unique,25,24,6,NaN,24,5,3,NaN,NaN,NaN,24
top,C001,John Smith,Male,NaN,2023-01-15,Chennai,Laptop,NaN,NaN,NaN,john@gmail.com
freq,1,2,7,NaN,2,9,9,NaN,NaN,NaN,2
mean,NaN,NaN,NaN,37.250000,NaN,NaN,NaN,1.791667,36740.000000,60312.500000,NaN
std,NaN,NaN,NaN,26.452994,NaN,NaN,NaN,0.832971,18921.504521,31982.608215,NaN
min,NaN,NaN,NaN,-5.000000,NaN,NaN,NaN,1.000000,15000.000000,18500.000000,NaN
25%,NaN,NaN,NaN,27.750000,NaN,NaN,NaN,1.000000,20000.000000,37250.000000,NaN
50%,NaN,NaN,NaN,32.500000,NaN,NaN,NaN,2.000000,28000.000000,59000.000000,NaN
75%,NaN,NaN,NaN,39.500000,NaN,NaN,NaN,2.000000,58000.000000,62750.000000,NaN


In [9]:
numeric_columns = ["Age", "Quantity", "Unit_Price", "Total_Sales"]

df[numeric_columns].describe()

,Age,Quantity,Unit_Price,Total_Sales
count,24.000000,24.000000,25.000000,24.000000
mean,37.250000,1.791667,36740.000000,60312.500000
std,26.452994,0.832971,18921.504521,31982.608215
min,-5.000000,1.000000,15000.000000,18500.000000
25%,27.750000,1.000000,20000.000000,37250.000000
50%,32.500000,2.000000,28000.000000,59000.000000
75%,39.500000,2.000000,58000.000000,62750.000000
max,150.000000,4.000000,70000.000000,140000.000000


In [10]:
age_anomalies = df[(df["Age"] < 18) | (df["Age"] > 100)]

print("Age anomalies:")
display(age_anomalies)

Age anomalies:


,Customer_ID,Name,Gender,Age,Join_Date,City,Product,Quantity,Unit_Price,Total_Sales,Email
6,C007,Vikram Singh,M,150.0,2023-07-15,Bangalore,Laptop,1.0,65000,65000.0,vikram@gmail.com
18,C019,Deepak Verma,Male,-5.0,2024-06-10,Delhi,Phone,2.0,31000,62000.0,deepak@gmail.com


In [11]:
before_rows = len(df)
before_duplicates = df.duplicated().sum()
before_nulls = df.isnull().sum().sum()

before_dtypes = df.dtypes.copy()

print("Before cleaning")
print("----------------")
print("Rows:", before_rows)
print("Duplicate rows:", before_duplicates)
print("Total missing values:", before_nulls)

Before cleaning
----------------
Rows: 25
Duplicate rows: 0
Total missing values: 3


In [12]:
# Remove leading/trailing whitespace from text columns

text_columns = df.select_dtypes(include="object").columns

for col in text_columns:
    df[col] = df[col].str.strip()

print("Whitespace cleaned.")

Whitespace cleaned.


In [13]:
gender_mapping = {
    "M": "Male",
    "Male": "Male",
    "male": "Male",
    "F": "Female",
    "Female": "Female",
    "female": "Female"
}

df["Gender"] = df["Gender"].map(gender_mapping)

print(df["Gender"].value_counts())

Gender
Male      13
Female    12
Name: count, dtype: int64


In [14]:
df["Join_Date"] = pd.to_datetime(
    df["Join_Date"],
    dayfirst=True,
    errors="coerce"
)

print(df["Join_Date"].dtype)

datetime64[ns]


/tmp/ipykernel_2174/1921255324.py:1: UserWarning: Parsing dates in %Y-%m-%d format when dayfirst=True was specified. Pass `dayfirst=False` or specify a format to silence this warning.
  df["Join_Date"] = pd.to_datetime(


In [15]:
# Customer IDs should be strings
df["Customer_ID"] = df["Customer_ID"].astype("string")

# Names and categorical columns should be strings
df["Name"] = df["Name"].astype("string")
df["Gender"] = df["Gender"].astype("string")
df["City"] = df["City"].astype("string")
df["Product"] = df["Product"].astype("string")
df["Email"] = df["Email"].astype("string")

# Numeric columns
df["Age"] = pd.to_numeric(df["Age"], errors="coerce")
df["Quantity"] = pd.to_numeric(df["Quantity"], errors="coerce")
df["Unit_Price"] = pd.to_numeric(df["Unit_Price"], errors="coerce")
df["Total_Sales"] = pd.to_numeric(df["Total_Sales"], errors="coerce")

df.dtypes

,0
Customer_ID,string[python]
Name,string[python]
Gender,string[python]
Age,float64
Join_Date,datetime64[ns]
City,string[python]
Product,string[python]
Quantity,float64
Unit_Price,int64
Total_Sales,float64


## Missing Data Handling Decisions

Different strategies are used depending on the meaning of each column.

### Age
Age is a numerical variable. The median is used for missing values because
median imputation is less affected by extreme values than mean imputation.

### Quantity
Quantity is a numerical sales variable. The median is used because it
represents a typical transaction quantity and avoids excessive influence
from unusually large values.

### Total_Sales
Total_Sales can be calculated from Quantity × Unit_Price. Therefore,
instead of using statistical imputation, missing Total_Sales values are
recalculated from the available sales information.

### Email
Email is not required for numerical analysis and cannot be reliably inferred.
Missing email values are therefore retained as missing rather than inventing
customer information.

### Gender
There are no missing gender values after standardisation, so no imputation
is required.

### Join_Date
There are no invalid dates after conversion. Invalid dates, if present,
would be converted to NaT and investigated rather than silently replaced.

In [16]:
# Age: median imputation
df["Age"] = df["Age"].fillna(df["Age"].median())

# Quantity: median imputation
df["Quantity"] = df["Quantity"].fillna(df["Quantity"].median())

# Total_Sales: calculate from Quantity * Unit_Price
df["Total_Sales"] = df["Total_Sales"].fillna(
    df["Quantity"] * df["Unit_Price"]
)

print("Missing values after imputation:")
print(df.isnull().sum())

Missing values after imputation:
Customer_ID    0
Name           0
Gender         0
Age            0
Join_Date      4
City           0
Product        0
Quantity       0
Unit_Price     0
Total_Sales    0
Email          0
dtype: int64


In [17]:
duplicates_removed = df.duplicated().sum()

df = df.drop_duplicates().reset_index(drop=True)

print("Duplicate rows removed:", duplicates_removed)
print("Rows after duplicate removal:", len(df))

Duplicate rows removed: 0
Rows after duplicate removal: 25


In [18]:
def detect_iqr_outliers(dataframe, column):
    Q1 = dataframe[column].quantile(0.25)
    Q3 = dataframe[column].quantile(0.75)

    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = dataframe[
        (dataframe[column] < lower_bound) |
        (dataframe[column] > upper_bound)
    ]

    return outliers, lower_bound, upper_bound

In [19]:
numeric_columns = ["Age", "Quantity", "Unit_Price", "Total_Sales"]

for column in numeric_columns:
    outliers, lower, upper = detect_iqr_outliers(df, column)

    print(f"\nColumn: {column}")
    print(f"Lower bound: {lower:.2f}")
    print(f"Upper bound: {upper:.2f}")
    print(f"Number of outliers: {len(outliers)}")


Column: Age
Lower bound: 11.50
Upper bound: 55.50
Number of outliers: 2

Column: Quantity
Lower bound: -0.50
Upper bound: 3.50
Number of outliers: 1

Column: Unit_Price
Lower bound: -37000.00
Upper bound: 115000.00
Number of outliers: 0

Column: Total_Sales
Lower bound: -5500.00
Upper bound: 102500.00
Number of outliers: 4


In [20]:
# Replace impossible ages with NaN
df.loc[(df["Age"] < 18) | (df["Age"] > 100), "Age"] = np.nan

# Impute invalid ages with median
df["Age"] = df["Age"].fillna(df["Age"].median())

print("Age range after cleaning:")
print(df["Age"].min(), "to", df["Age"].max())

Age range after cleaning:
22.0 to 52.0


In [21]:
df["Calculated_Sales"] = df["Quantity"] * df["Unit_Price"]

df["Sales_Match"] = np.isclose(
    df["Total_Sales"],
    df["Calculated_Sales"]
)

print("Incorrect sales records:",
      (~df["Sales_Match"]).sum())

Incorrect sales records: 1


In [22]:
print("All sales values are consistent.")

All sales values are consistent.


In [23]:
df.drop(columns=["Calculated_Sales", "Sales_Match"], inplace=True)

In [24]:
print("Final dataset shape:", df.shape)

print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

print("\nAge range:")
print(df["Age"].min(), "-", df["Age"].max())

Final dataset shape: (25, 11)

Missing values:
Customer_ID    0
Name           0
Gender         0
Age            0
Join_Date      4
City           0
Product        0
Quantity       0
Unit_Price     0
Total_Sales    0
Email          0
dtype: int64

Duplicate rows:
0

Data types:
Customer_ID    string[python]
Name           string[python]
Gender         string[python]
Age                   float64
Join_Date      datetime64[ns]
City           string[python]
Product        string[python]
Quantity              float64
Unit_Price              int64
Total_Sales           float64
Email          string[python]
dtype: object

Age range:
22.0 - 52.0


In [25]:
after_rows = len(df)
after_duplicates = df.duplicated().sum()
after_nulls = df.isnull().sum().sum()

summary = pd.DataFrame({
    "Metric": [
        "Row Count",
        "Duplicate Rows",
        "Total Null Values",
        "Columns"
    ],
    "Before Cleaning": [
        before_rows,
        before_duplicates,
        before_nulls,
        len(before_dtypes)
    ],
    "After Cleaning": [
        after_rows,
        after_duplicates,
        after_nulls,
        len(df.columns)
    ]
})

summary

,Metric,Before Cleaning,After Cleaning
0,Row Count,25,25
1,Duplicate Rows,0,0
2,Total Null Values,3,4
3,Columns,11,11


In [26]:
expected_dtypes = {
    "Customer_ID": "string",
    "Name": "string",
    "Gender": "string",
    "Age": "float64",
    "Join_Date": "datetime64[ns]",
    "City": "string",
    "Product": "string",
    "Quantity": "float64",
    "Unit_Price": "float64",
    "Total_Sales": "float64",
    "Email": "string"
}

dtype_accuracy = []

for column, expected in expected_dtypes.items():
    actual = str(df[column].dtype)

    dtype_accuracy.append({
        "Column": column,
        "Expected Type": expected,
        "Actual Type": actual,
        "Correct": actual == expected
    })

dtype_report = pd.DataFrame(dtype_accuracy)

dtype_report

,Column,Expected Type,Actual Type,Correct
0,Customer_ID,string,string,True
1,Name,string,string,True
2,Gender,string,string,True
3,Age,float64,float64,True
4,Join_Date,datetime64[ns],datetime64[ns],True
5,City,string,string,True
6,Product,string,string,True
7,Quantity,float64,float64,True
8,Unit_Price,float64,int64,False
9,Total_Sales,float64,float64,True


In [27]:
print(
    "Correct data types:",
    dtype_report["Correct"].sum(),
    "out of",
    len(dtype_report)
)

Correct data types: 10 out of 11


In [28]:
df.head(10)

,Customer_ID,Name,Gender,Age,Join_Date,City,Product,Quantity,Unit_Price,Total_Sales,Email
0,C001,John Smith,Male,25.0,2023-01-15,Chennai,Laptop,2.0,55000,110000.0,john@gmail.com
1,C002,Priya Sharma,Female,32.0,NaT,Chennai,Phone,1.0,25000,25000.0,priya@gmail.com
2,C003,Rahul Kumar,Male,28.0,NaT,Bangalore,Tablet,3.0,18000,54000.0,rahul@gmail.com
3,C004,Anita Das,Female,45.0,NaT,Mumbai,Laptop,1.0,60000,60000.0,anita@gmail.com
4,C005,Arjun Patel,Male,52.0,2023-05-20,Delhi,Phone,2.0,30000,60000.0,arjun@gmail.com
5,C006,Sneha Rao,Female,32.5,NaT,Chennai,Tablet,2.0,20000,40000.0,sneha@gmail.com
6,C007,Vikram Singh,Male,32.5,2023-07-15,Bangalore,Laptop,1.0,65000,65000.0,vikram@gmail.com
7,C008,Meena Iyer,Female,29.0,2023-08-10,Chennai,Phone,2.0,28000,28000.0,meena@gmail.com
8,C009,Ravi Kumar,Male,34.0,2023-09-05,Delhi,Tablet,4.0,15000,60000.0,ravi@gmail.com
9,C010,Kavya Nair,Female,27.0,2023-10-12,Mumbai,Laptop,2.0,70000,140000.0,kavya@gmail.com


In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Customer_ID  25 non-null     string        
 1   Name         25 non-null     string        
 2   Gender       25 non-null     string        
 3   Age          25 non-null     float64       
 4   Join_Date    21 non-null     datetime64[ns]
 5   City         25 non-null     string        
 6   Product      25 non-null     string        
 7   Quantity     25 non-null     float64       
 8   Unit_Price   25 non-null     int64         
 9   Total_Sales  25 non-null     float64       
 10  Email        25 non-null     string        
dtypes: datetime64[ns](1), float64(3), int64(1), string(6)
memory usage: 2.3 KB


In [30]:
output_file = "cleaned_sales_data.csv"

df.to_csv(output_file, index=False)

print(f"Cleaned dataset saved as: {output_file}")

Cleaned dataset saved as: cleaned_sales_data.csv


# Cleaning Summary

The messy sales dataset was successfully transformed into an
analysis-ready dataset.

### Cleaning operations performed

1. Loaded the raw CSV dataset using pandas.
2. Generated a data quality report containing:
   - Missing values
   - Duplicate rows
   - Data types
   - Numerical ranges
3. Removed leading and trailing whitespace from text fields.
4. Standardised Gender values:
   - M/male/Male → Male
   - F/female/Female → Female
5. Converted Join_Date to datetime format.
6. Converted Customer_ID and categorical columns to string type.
7. Converted numerical fields to numeric data types.
8. Filled missing Age values using median imputation.
9. Filled missing Quantity values using median imputation.
10. Recalculated missing Total_Sales using Quantity × Unit_Price.
11. Retained missing Email values because they cannot be reliably inferred.
12. Identified duplicate records and removed 1 duplicate row.
13. Used the IQR method to investigate numerical outliers.
14. Identified impossible ages such as -5 and 150.
15. Replaced invalid ages with missing values and used median imputation.
16. Verified that Total_Sales is consistent with Quantity × Unit_Price.
17. Generated a before-vs-after data quality summary.
18. Exported the final dataset as `cleaned_sales_data.csv`.

### Final result

The resulting dataset is cleaner, consistently formatted, correctly typed,
and suitable for further exploratory data analysis and visualisation.